### PR TITLE
P#57341 Search in view tables

### PR DESCRIPTION
**Pull request by tung-le-esg**
P#57341 Search in view tables

### CODE REVIEW

File: plugin/Gui/AdminPage.php [hunk 1]
```diff
@@ -43,4 +43,12 @@ public function registerForms()
 			$pInputModelRenderer->registerFields($pFormModel);
 		}
 	}
+
+	public function generateSearchForm($page,$button)
+	{
+		echo '<form action="'.admin_url('admin.php').'" method="get" id="onoffice-form-search">'.$page.
```
Comment (jayay) [inline plugin/Gui/AdminPage.php:49]: Please remember to escape values

File: plugin/Gui/AdminPageFormList.php [hunk 2]
```diff
@@ -96,6 +96,10 @@ public function renderContent()
 		$this->generatePageMainTitle(__('Forms', 'onoffice-for-wp-websites'));
 
 		$this->_pFormsTable->prepare_items();
+		$page = '<input type="hidden" id="fname" name="type" value="'.$_GET['type'].'">
```
Comment (jayay) [inline plugin/Gui/AdminPageFormList.php:99]: Please add escaping here as well

File: plugin/Record/RecordManagerReadForm.php [hunk 3]
```diff
@@ -86,6 +86,10 @@ public function getRecordsSortedAlphabetically()
         $columns = implode(', ', $this->getColumns());
         $join = implode("\n", $this->getJoins());
         $where = "(".implode(") AND (", $this->getWhere()).")";
+		if (!empty($_GET["search"]))
+		{
+			$where .= "AND name LIKE '%".$_GET['search']."%' OR template LIKE '%".$_GET['search']."%' OR recipient LIKE '%".$_GET['search']."%'";
```
Comment (jayay) [inline plugin/Record/RecordManagerReadForm.php:91]: Please add SQL escaping here. This is very important

File: plugin/Record/RecordManagerReadListViewAddress.php [hunk 4]
```diff
@@ -82,13 +82,16 @@ public function getRecordsSortedAlphabetically()
         $columns = implode(', ', $this->getColumns());
         $join = implode("\n", $this->getJoins());
         $where = "(".implode(") AND (", $this->getWhere()).")";
+        if (!empty($_GET["search"]))
+		{
+			$where .= "AND name LIKE '%".$_GET['search']."%' OR template LIKE '%".$_GET['search']."%'";
```
Comment (jayay) [inline plugin/Record/RecordManagerReadListViewAddress.php:87]: Same thing about escaping.

File: plugin/Record/RecordManagerReadListViewEstate.php [hunk 5]
```diff
@@ -87,12 +87,17 @@ public function getRecordsSortedAlphabetically()
         $columns = implode(', ', $this->getColumns());
         $join = implode("\n", $this->getJoins());
         $where = "(".implode(") AND (", $this->getWhere()).")";
+		if (!empty($_GET["search"]))
+		{
+			$where .= "AND name LIKE '%".$_GET['search']."%' OR template LIKE '%".$_GET['search']."%'";
```
Comment (jayay) [inline plugin/Record/RecordManagerReadListViewEstate.php:92]: And here as well

**Review by jayay — CHANGES_REQUESTED** [review]

File: plugin/Gui/AdminPage.php (hunk 1 above)
Comment (vu-vuong-esg) [reply plugin/Gui/AdminPage.php]: @jayay $page is an input tag. I have already escape the value of this input before.

**Review by jayay — CHANGES_REQUESTED** [review]
I have a couple of concerns regarding HTML escaping. Please also check the indentation, tabs and  spaces might be mixed.

**Review by jayay — COMMENTED** [review]
`$type !== null` would be a little bit nicer in AdminPage.php. Other than that I would kindly ask you to please check the indents, tabs and spaces seem to be mixed.

**Comment by jayay** [comment]
Please don't use `Closure::bind()` in tests anymore. Its usage doesn't reflect real-life usage and leads to complicated, inaccurate tests.

**Comment by vu-vuong-esg** [comment]
> Please don't use `Closure::bind()` in tests anymore. Its usage doesn't reflect real-life usage and leads to complicated, inaccurate tests.

i fixed it, can you check it ?

**Comment by jayay** [comment]
Looks good so far. Strict types in parameters and method returns would be nice, as well as tests that test if the `RecordManager` classes work correctly.

File: tests/TestClassRecordManagerReadListViewAddress.php [hunk 6]
```diff
@@ -22,8 +22,10 @@
 namespace onOffice\tests;
 
 use Closure;
+use DI\ContainerBuilder;
 use onOffice\WPlugin\Record\RecordManagerReadListViewAddress;
 use WP_UnitTestCase;
+use function cli\confirm;
```
Comment (jayay) [inline tests/TestClassRecordManagerReadListViewAddress.php:28]: this looks odd

**Review by jayay — COMMENTED** [review]
Looks good! There's just one small thing (the line `use function cli\confirm;`)

File: tests/TestClassRecordManagerReadListViewAddress.php (hunk 6 above)
Comment (tang-hien-egs) [reply tests/TestClassRecordManagerReadListViewAddress.php]: Sorry, i fixed it, can you check PR?

**Review by jayay — DISMISSED** [review]

**Comment by michaelarnold-dev** [comment]
Test with version 2.21.4-23-ga61350b2 on https://www.227.wp.myonoffice.com/.

1.) The button has black font, not blue as in Backend_Listview_Forms-search-desired.jpg.
2.) Unit lists are not searchable. This was not explicitly listed in the specifications. But this should still be implemented.
3.) Requirement: When searching for a form, the query should be looked up in the form's title, email subject, email address, and chosen template -> Email subject is not searched.
4.) After searching, unit views are also listed in the estate list. But these should not be displayed in the normal estate list.
Backend_Listview_Forms-search-current (1)
Backend_Listview_Forms-search-desired

**Comment by tang-hien-egs** [comment]
> Test with version 2.21.4-23-ga61350b2 on https://www.227.wp.myonoffice.com/.
> 
> 1.) The button has black font, not blue as in Backend_Listview_Forms-search-desired.jpg. 2.) Unit lists are not searchable. This was not explicitly listed in the specifications. But this should still be implemented. 3.) Requirement: When searching for a form, the query should be looked up in the form's title, email subject, email address, and chosen template -> Email subject is not searched. 4.) After searching, unit views are also listed in the estate list. But these should not be displayed in the normal estate list.
> 
> current Backend_Listview_Forms-search-current (1) desired Backend_Listview_Forms-search-desired

i fixed it and recorded video like link :https://files.fm/f/tzw5bakcf, can you check PR ?

**Comment by tang-hien-egs** [comment]
> 

@michaelarnold-dev , i fixed it and record video like link : https://files.fm/f/suk88qzgu

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs At addresses is still a small GUI error, GUI elements overlap. See screenshot. The search button and the search field should be moved down a bit

.

**Comment by tang-hien-egs** [comment]
> @tang-hien-egs At addresses is still a small GUI error, GUI elements overlap. See screenshot. The search button and the search field should be moved down a bit  .

@michaelarnold-dev , i fixed it, can you check help me ?

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs Thank you for fixing it. I guess @jayay needs to review it before we can test it again.

**Review by jayay — APPROVED** [review]

**Pull request merged**